src/lightning/pytorch/loops/training_epoch_loop.py

```diff
@@ -189,22 +189,30 @@ def advance(self, data_fetcher: _DataFetcher) -> None:
         # we are going to train first so the val loop does not need to restart
         self.val_loop.restarting = False
 
-        batch_idx = data_fetcher.fetched if isinstance(data_fetcher, _DataLoaderIterDataFetcher) else self.batch_idx + 1
-        batch = next(data_fetcher)
+        if using_dataloader_iter := isinstance(data_fetcher, _DataLoaderIterDataFetcher):
+            dataloader_iter = next(data_fetcher)
+            # hook's batch_idx and dataloader_idx arguments correctness cannot be guaranteed in this setting
+            batch = data_fetcher._batch
+            batch_idx = data_fetcher._batch_idx
+        else:
+            dataloader_iter = None
+            batch, _, __ = next(data_fetcher)
+            # TODO: we should instead use the batch_idx returned by the fetcher, however, that will require saving the
+            # fetcher state so that the batch_idx is correct after restarting
+            batch_idx = self.batch_idx + 1
```

Comment:
this seems to be very repetitive, cam we wrap it as private func?

Reply:
The logic is slightly different in each instance. I don't see how it could be reused at the moment. Perhaps in the future after the TODO is resolved